### PR TITLE
Make explicit arguments for context bounds an error from 3.5

### PR DESCRIPTION
**Pull request by odersky**
Warning in 3.4. Rewrite available in 3.5-migration.

[test_scala2_library_tasty]

### CODE REVIEW

**Comment by odersky** [comment]
There's a difference in neg/hidden-type-errors and a failure in ReplTests. Both look like some internals of quotes pass
an explicit argument to a context bound parameter. It should be a `using` argument instead. @nicolasstucki , can you check?

**Comment by odersky** [comment]
@nicolasstucki There was an error in `scala2-library-tasty-test`. Can you advise what the reason could be?
```scala
- Error: /__w/dotty/dotty/scala2-library-tasty-tests/src/Main.scala:7:19 ------
Error:  7 |  case Red, Green, Blue
Error:    |                   ^
Error:    |Context bounds will map to context parameters.
Error:    |A `using` clause is needed to pass explicit arguments to them.
Error:    |This code can be rewritten automatically under -rewrite -source 3.5-migration.
Error:  one error found
```

**Comment by nicolasstucki** [comment]
> @nicolasstucki There was an error in `scala2-library-tasty-test`. Can you advise what the reason could be?
> 
> ```scala
> - Error: /__w/dotty/dotty/scala2-library-tasty-tests/src/Main.scala:7:19 ------
> Error:  7 |  case Red, Green, Blue
> Error:    |                   ^
> Error:    |Context bounds will map to context parameters.
> Error:    |A `using` clause is needed to pass explicit arguments to them.
> Error:    |This code can be rewritten automatically under -rewrite -source 3.5-migration.
> Error:  one error found
> ```

The reason is that we compile the standard library with this new mode. In particular the `def apply[T: ClassTag](xs: T*): Array[T]` of `Array` now expects the `using` when the `ClassTag` argument is passed explicitly. This is happening in the desugaring of the enumeration in:

```scala
    private[this] val $values: Array[Color] =
      Array.apply[Color]([this.Red,this.Green,this.Blue : Color]*)(
        scala.reflect.ClassTag.apply[Color](classOf[Color]))
```

The solution is to make sure all desugarings that explicitly apply a context-bound set the `ApplyKind` to `Using`.

This particular error can be fixed in `DesugarEnums.scala`
```diff
private def ArrayLiteral(values: List[Tree], tpt: Tree)(using Context): Tree =
    val clazzOf = TypeApply(ref(defn.Predef_classOf.termRef), tpt :: Nil)
    val ctag    = Apply(TypeApply(ref(defn.ClassTagModule_apply.termRef), tpt :: Nil), clazzOf :: Nil)
    val apply   = Select(ref(defn.ArrayModule.termRef), nme.apply)
-   Apply(Apply(TypeApply(apply, tpt :: Nil), values), ctag :: Nil)
+   Apply(Apply(TypeApply(apply, tpt :: Nil), values), ctag :: Nil).setApplyKind(ApplyKind.Using)
```

This fix makes `scala2-library-tasty-tests/run` and `scala2-library-tasty-tests/test` pass, though we might find other cases when running all tests with the TASTy library in `test_scala2_library_tasty` CI job. This one is usually tested on nightlies, I will enable it on this PR to make sure we are not missing something else.

**Comment by odersky** [comment]
@nicolasstucki Ready for review.

File: tests/neg/context-bounds-migration.scala
Comment (nicolasstucki) [inline tests/neg/context-bounds-migration.scala:1]: Tests with warnings should now be added to `tests/warn`.


`tests/warn/context-bounds-migration.scala`
```scala
class C[T]
def foo[X: C] = ()

given [T]: C[T] = C[T]()

def Test =
  foo(C[Int]())  // warn
  foo(using C[Int]()) // ok
```

File: compiler/src/dotty/tools/dotc/typer/Migrations.scala [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      case Block(mdef @ DefDef(_, vparams :: Nil, _, _) :: Nil, _: Closure) =>
+        val arity = vparams.length
+        if (arity > 0) ("", "") else ("(() => ", "())")
+      case _ =>
+        ("(() => ", ")")
+    }
+    def remedy =
+      if ((prefix ++ suffix).isEmpty) "simply leave out the trailing ` _`"
+      else s"use `$prefix<function>$suffix` instead"
+    def rewrite = Message.rewriteNotice("This construct", `3.4-migration`)
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Migrations.scala:75]: Maybe `rewriteNotice` should receive a `MigrationVersion` instead. I will check that later an refactor all the use cases if needed.

File: compiler/src/dotty/tools/dotc/typer/Migrations.scala [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import NameKinds.ContextBoundParamName
+import rewrites.Rewrites.patch
+import util.Spans.Span
+import rewrites.Rewrites
+
+/** A utility trait containing source-dependent deprecation messages
+ *  and migrations.
+ */
+trait Migrations:
+  this: Typer =>
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Migrations.scala:29]: I am not convinced we should make this one a mixin. The one use case right now for `typedAsFunction`, all other functions that are here and could go in here seem to be static.

I would suggest making `Migrations` an `object` and redefining `Typer.typedAsFunction` as
```scala
def typedAsFunction(tree: untpd.PostfixOp, pt: Type)(using Context): Tree =
    migrate(Migrations.typedAsFunction(tree, pt), throw new AssertionError("can't retype a PostfixOp"))
```

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 3]
```diff
@@ -158,6 +158,9 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
   // Overridden in derived typers
   def newLikeThis(nestingLevel: Int): Typer = new Typer(nestingLevel)
 
+  // Overridden to do nothing in derived typers
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:161]:
```suggestion
  // Overridden to call `disabled` in derived typers
```

**Review by nicolasstucki — APPROVED** [review]

**Pull request merged**